### PR TITLE
add roaring_bitmap_rank_many(): get `rank()` values in Bulk

### PR DESCRIPTION
**Pull request by longqimin**
this pr provide an api implementation for fast getting `rank` of many elements in bulk mode.
as well as test/benchmark.
```
$ microbenchmarks/bench --benchmark_filter="RankMany*"
2023-11-29T09:55:50+08:00
Running microbenchmarks/bench
Run on (36 X 4600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x18)
  L1 Instruction 32 KiB (x18)
  L2 Unified 1024 KiB (x18)
  L3 Unified 25344 KiB (x1)
Load Average: 0.57, 0.73, 0.61
AVX-2 hardware: yes
AVX-512: supported by compiler
AVX-512 hardware: no
In RAM volume in MiB (estimated): 1.802828
benchmarking other files: You may pass is a data directory as a parameter.
data source: CRoaring/benchmarks/realdata/census1881
number of bitmaps: 200
x64: detected
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-----------------------------------------------------------------------
Benchmark             Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------
RankManySlow      14070 ns        14030 ns        49497 GHz=4.60848 cycles=61.537k instructions=145.324k
RankMany           5616 ns         5601 ns       125617 GHz=4.42231 cycles=24.734k instructions=75.405k
```

### CODE REVIEW

**Comment by longqimin** [comment]
@lemire Would you take a look ?

**Comment by lemire** [comment]
Of course I will review.

File: include/roaring/roaring.h [hunk 1]
```diff
@@ -815,6 +815,15 @@ bool roaring_bitmap_select(const roaring_bitmap_t *r, uint32_t rank,
  */
 uint64_t roaring_bitmap_rank(const roaring_bitmap_t *r, uint32_t x);
 
+/**
+ * roaring_bitmap_rank_many is an `Bulk` version of `roaring_bitmap_rank`
+ * it puts rank value of each element in `[begin .. end)` to `ans[]`
+ *
+ * the values in `[begin .. end)` must be sorted in Ascending order;
+ * the `ans` must have enough size.
```
Comment (lemire) [inline include/roaring/roaring.h:823]: May be we can be more precise and state what is meant by _size_. It is `size_t length = (end-begin)`. Correct?

**Comment by lemire** [comment]
@longqimin This looks *very* good to me. Can you check my comment above?

File: include/roaring/containers/array.h [hunk 2]
```diff
@@ -412,6 +412,29 @@ inline int array_container_rank(const array_container_t *arr, uint16_t x) {
...
+inline uint32_t array_container_rank_many(const array_container_t *arr, uint64_t start_rank, const uint32_t* begin, const uint32_t* end, uint64_t* ans){
+    const uint16_t high = (*begin) >> 16;
+    uint32_t pos = 0;
+    const uint32_t* iter = begin;
+    for(; iter != end; iter++) {
+        uint32_t x = *iter;
+        uint16_t xhigh = x >> 16;
+        if(xhigh != high) return iter - begin;// stop at next container
+
+        const int32_t idx = binarySearch(arr->array+pos, arr->cardinality-pos, x&0xFFFF);
```
Comment (lemire) [inline include/roaring/containers/array.h:425]:
```suggestion
        const int32_t idx = binarySearch(arr->array+pos, arr->cardinality-pos, (uint16_t)x);
```

File: include/roaring/containers/array.h [hunk 3]
```diff
@@ -412,6 +412,29 @@ inline int array_container_rank(const array_container_t *arr, uint16_t x) {
...
 }
 
+// bulk version of array_container_rank(); return number of consumed elements
+inline uint32_t array_container_rank_many(const array_container_t *arr, uint64_t start_rank, const uint32_t* begin, const uint32_t* end, uint64_t* ans){
+    const uint16_t high = (*begin) >> 16;
+    uint32_t pos = 0;
+    const uint32_t* iter = begin;
+    for(; iter != end; iter++) {
+        uint32_t x = *iter;
+        uint16_t xhigh = x >> 16;
```
Comment (lemire) [inline include/roaring/containers/array.h:422]:
```suggestion
        uint16_t xhigh = (uint16_t)(x >> 16);
```

File: src/containers/bitset.c [hunk 4]
```diff
@@ -1232,6 +1232,29 @@ int bitset_container_rank(const bitset_container_t *container, uint16_t x) {
   return sum;
 }
 
+uint32_t bitset_container_rank_many(const bitset_container_t *container, uint64_t start_rank, const uint32_t* begin, const uint32_t* end, uint64_t* ans){
+  const uint16_t high = (*begin) >> 16;
```
Comment (lemire) [inline src/containers/bitset.c:1236]:
```suggestion
  const uint16_t high = (uint16_t)((*begin) >> 16);
```

File: src/containers/bitset.c [hunk 5]
```diff
@@ -1232,6 +1232,29 @@ int bitset_container_rank(const bitset_container_t *container, uint16_t x) {
...
+  int i = 0;
+  int sum = 0;
+  const uint32_t* iter = begin;
+  for(; iter != end; iter++) {
+      uint32_t x = *iter;
+      uint16_t xhigh = x >> 16;
```
Comment (lemire) [inline src/containers/bitset.c:1242]:
```suggestion
      uint16_t xhigh = (uint16_t)(x >> 16);
```

File: src/containers/bitset.c [hunk 6]
```diff
@@ -1232,6 +1232,29 @@ int bitset_container_rank(const bitset_container_t *container, uint16_t x) {
...
+      if(xhigh != high) return iter - begin; // stop at next container
+
+      uint16_t xlow = x & 0xFFFF;
```
Comment (lemire) [inline src/containers/bitset.c:1245]:
```suggestion
      uint16_t xlow = (uint16_t)x;
```

File: src/containers/run.c [hunk 7]
```diff
@@ -883,6 +883,39 @@ int run_container_rank(const run_container_t *container, uint16_t x) {
...
     return sum;
 }
+uint32_t run_container_rank_many(const run_container_t *container, uint64_t start_rank, const uint32_t* begin, const uint32_t* end, uint64_t* ans){
+    const uint16_t high = (*begin) >> 16;
+    const uint32_t* iter = begin;
+    int sum = 0;
+    int i = 0;
+    for(;iter != end; iter++) {
+        uint32_t x = *iter;
+        uint16_t xhigh = x >> 16;
```
Comment (lemire) [inline src/containers/run.c:893]:
```suggestion
        uint16_t xhigh = (uint16_t)(x >> 16);
```

File: include/roaring/roaring.h (hunk 1 above)
Comment (longqimin) [reply include/roaring/roaring.h:823]: ok. i borrowed a doc from `roaring_bitmap_to_uint32_array()`. 
```
 *     ans = malloc((end-begin) * sizeof(uint64_t));

File: include/roaring/containers/array.h (hunk 2 above)
Comment (longqimin) [reply include/roaring/containers/array.h]: done

File: include/roaring/containers/array.h [hunk 8]
```diff
@@ -412,6 +412,29 @@ inline int array_container_rank(const array_container_t *arr, uint16_t x) {
     }
 }
 
+// bulk version of array_container_rank(); return number of consumed elements
+inline uint32_t array_container_rank_many(const array_container_t *arr, uint64_t start_rank, const uint32_t* begin, const uint32_t* end, uint64_t* ans){
+    const uint16_t high = (*begin) >> 16;
```
[the remark this answers is not in the document]
Comment (longqimin) [reply include/roaring/containers/array.h]: done

File: include/roaring/containers/array.h (hunk 3 above)
Comment (longqimin) [reply include/roaring/containers/array.h]: done

File: src/containers/bitset.c (hunk 4 above)
Comment (longqimin) [reply src/containers/bitset.c]: done

File: src/containers/bitset.c (hunk 5 above)
Comment (longqimin) [reply src/containers/bitset.c]: done

File: src/containers/bitset.c (hunk 6 above)
Comment (longqimin) [reply src/containers/bitset.c]: done

File: src/containers/run.c (hunk 7 above)
Comment (longqimin) [reply src/containers/run.c]: done

**Comment by lemire** [comment]
Merged. Will be part of the next release. Thanks.

**Comment by Dr-Emann** [comment]
> the values in `[begin .. end)` must be sorted in Ascending order

It might be worth documenting which of the following can occur if this is not upheld:

* Incorrect values
* Infinite looping
* Full Undefined Behavior

**Pull request merged**